File: frontend/src/types/envelopeType.ts
export type Direction = "c2s" | "s2c";
export type Actor = "assistant" | "coder" | "writer";
export type Action = "stream";
export type Modifier = "start" | "chunk" | "end";

export type Envelope<T = unknown> = {
  // protocol
  v: "1";

  // identity & timing
  id: string; // message id (uuid)
  ts: number; // Unix timestamp in milliseconds

  // correlation
  requestId?: string; // request you're replying to
  streamId?: string; // server-minted for streams
  seq?: number; // per-stream sequence, starting at 0

  // control
  direction: Direction;
  actor: Actor;
  action: Action;
  modifier: Modifier;

  // payload
  data: T;

  // errors
  error?: {
    code: ErrorCode; // see taxonomy below
    message: string;
    details?: unknown;
  };
};

export type ErrorCode =
  | "E_INVALID"
  | "E_UNAUTHORIZED"
  | "E_FORBIDDEN"
  | "E_NOT_FOUND"
  | "E_CONFLICT"
  | "E_RATE_LIMITED"
  | "E_TIMEOUT"
  | "E_OVERFLOW"
  | "E_UNAVAILABLE"
  | "E_INTERNAL";

File: backend/core/sockets/envelope_type.py
import uuid
from datetime import datetime, timezone
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from core.sockets.types import Message


class AliasedBaseModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    def model_dump_json(self, **kwargs) -> str:
        return super().model_dump_json(by_alias=True, **kwargs)


class ErrorDetails(AliasedBaseModel):
    code: Literal[
        "E_INVALID",
        "E_UNAUTHORIZED",
        "E_FORBIDDEN",
        "E_NOT_FOUND",
        "E_CONFLICT",
        "E_RATE_LIMITED",
        "E_TIMEOUT",
        "E_OVERFLOW",
        "E_UNAVAILABLE",
        "E_INTERNAL",
    ]
    message: str
    details: dict | None = None


Direction = Literal["c2s", "s2c"]
Actor = Literal["assistant", "coder", "writer"]
Action = Literal["stream"]
Modifier = Literal["start", "chunk", "end"]


class Envelope(AliasedBaseModel):
    # protocol
    v: str = "1"

    # identity & timing
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    ts: int = Field(
        default_factory=lambda: int(datetime.now(timezone.utc).timestamp() * 1000)
    )

    # correlation
    request_id: str | None = None  # requestId in TS
    stream_id: str | None = None  # streamId in TS
    seq: int | None = None  # per-stream sequence, starting at 1

    # control
    direction: Direction
    actor: Actor
    action: Action
    modifier: Modifier

    # payload
    data: dict | list[Message]

    # errors
    error: ErrorDetails | None = None


class AckOk(AliasedBaseModel):
    ok: bool = True
    request_id: str
    stream_id: str


class Error(AliasedBaseModel):
    code: str
    message: str


class AckFail(AliasedBaseModel):
    ok: bool = False
    error: Error
